engine/consensus/compliance/engine.go

```diff
@@ -66,49 +54,118 @@ func NewEngine(
 	me module.Local,
 	prov network.Engine,
 	core *Core) (*Engine, error) {
-	e := &Engine{
-		unit:             engine.NewUnit(),
-		log:              log.With().Str("compliance", "engine").Logger(),
-		me:               me,
-		mempool:          core.mempool,
-		metrics:          core.metrics,
-		headers:          core.headers,
-		payloads:         core.payloads,
-		state:            core.state,
-		tracer:           core.tracer,
-		prov:             prov,
-		core:             core,
-		pendingEventSink: make(EventSink),
-		blockSink:        make(EventSink),
-		voteSink:         make(EventSink),
-	}
 
-	var err error
-	// register the core with the network layer and store the conduit
-	e.con, err = net.Register(engine.ConsensusCommittee, e)
-	if err != nil {
-		return nil, fmt.Errorf("could not register core: %w", err)
-	}
+	notify := make(chan struct{})
 
 	// FIFO queue for block proposals
-	e.pendingBlocks, err = fifoqueue.NewFifoQueue(
+	pendingBlocks, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultBlockQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound receipts: %w", err)
 	}
 
 	// FIFO queue for block votes
-	e.pendingVotes, err = fifoqueue.NewFifoQueue(
+	pendingVotes, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultVoteQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound approvals: %w", err)
 	}
 
-	return e, nil
+	// define message queueing behaviour
+	handler := engine.NewMessageHandler(
```

Comment:
In my mind, the `MessageHandler`'s core contribution is to encapsulate the `notify` channel and the `DoNotify` method. 
If feel that all the pattern logic is very verbose. An implementation as a switch statement in the engine's `Process` method, could be much more concise. Lets imagine, we have a `Poker`, that only comprises the  `notify channel` with a  [`DoNotify`](https://github.com/onflow/flow-go/blob/2335b4a9b607d0a977c9bb1deb59fee3f42b8607/engine/enqueue.go#L95-L100) method:

```golang
	switch event.Msg.(type) {
	case *events.SyncedBlock:
		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageSyncedBlock)
		// a block that is synced has to come locally, from the synchronization engine
		// the block itself will contain the proposer to indicate who created it
		if event.OriginID != e.me.NodeID() {
			log.Warn().Msgf("synced block with non-local origin (local: %x, origin: %x)", e.me.NodeID(), event.OriginID)
			return
		}
		e.pendingBlocks.Push(event)
		e.Poker.DoNotify()
	case *messages.BlockProposal:
		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockProposal)
		e.pendingBlocks.Push(event)
		e.Poker.DoNotify()
	case *messages.BlockVote:
		e.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockVote)
		e.pendingVotes.Push(event)
		e.Poker.DoNotify()
	}
```
In my mind, this would remove lots of indirection.

Reply:
I agree it is more verbose, the benefit for me is to provide a common touchpoint for things like metrics and tracing. Rather than manually (and verbosely) capture metrics separately in every engine for every message, we can encapsulate this in the `MessageHandler`. The way it's implemented now, the engine still has to specify metric types, but at least the `Pattern` constructor can enforce that a metric is set for every message. Or better, the message types themselves contain their metric label (and this is enforced by a `Message` interface similar to the `Entity` interface). Tracing the path of a message through the system could similarly use this as a consistent touchpoint.

The other thing I was considering with this was having the `MessageProcessor` manage the worker go-routines for the engine, rather than re-implementing what is likely to be a similar set of synchronization rules between the different inbound message types for each engine. Most engines either process everything individually (one global worker), or process a message type individually (one worker per message type), or process a message type with unlimited parallelism (`n` workers for that message type). It's definitely not clear that there's enough commonality to extract this, but worth keeping in mind.

Reply:
IMO, the use of `Pattern` is a good abstraction to me.

The abstraction basically says:
Given a stream of incoming messages, if you tell me how to filter messages (pattern matching), and where to store those messages, then I will store those filter messages, and provide you a channel to subscribe whenever a new messages is stored.

It becomes easy to think that if we provide the pattern matching rules, then it returns a channel to get notified.

I would even suggest to let `NewMessageHandler` return the `notify` channel, so you will only get a channel after providing me patterns. We can do so, because the `notify` channel is always unbuffered. Why? because the internal `doNotify` doesn't want to get blocked in case there is no handler subscribing the channel. But as the user of message handler, I don't want to expose the `DoNotify` method, because I would have to understand what is that for, and run into risk of forgetting to call it.